### PR TITLE
Purchases: Add `state.purchases.items` reducer

### PR DESCRIPTION
**Pull request by drewblaisdell**
Addresses part of https://github.com/Automattic/wp-calypso/issues/5046.

This PR copies the reducer and its tests from `lib/purchases/reducer` into `state/purchases/` and updates them such that:

- The tests test the reducer, instead of a Flux singleton.
- The reducer is exported in a function assigned as `items`.

It also sorts `state/action-types`, which was not entirely alphabetized.

This PR does not remove the old reducer or Flux store. Nor does it import the newly updated reducer into the combined global reducer. This will happen in a separate PR to keep the PR size manageable.

**Testing**
Apart from the changes to `action-types`, the changes here won't affect the actual build, so this only needs a code review. There are updated tests for the reducer that you can run with:

```bash
npm run test-client client/state/purchases/
```

Although the CI indicates that the tests have passed already.

- [ ] Code

Test live: https://calypso.live/?branch=update/purchases-reducer

### CODE REVIEW

File: client/state/purchases/index.js [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+ */
+import assign from 'lodash/assign';
+import find from 'lodash/find';
+import matches from 'lodash/matches';
+
+/**
+ * Internal Dependencies
+ */
+import {
+	PURCHASES_REMOVE,
```
Comment (scruffian) [inline client/state/purchases/index.js]: I guess we should probably alphabetize these.

File: client/state/purchases/index.js [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+	return purchases;
+}
+
+export const items = ( state = initialState, action ) => {
+	const { type } = action;
+
+	switch ( type ) {
+		case PURCHASES_REMOVE:
+			return assign( {}, state, {
```
Comment (scruffian) [inline client/state/purchases/index.js]: @gziolo says we can use spread operators in Calypso now :)

File: client/state/purchases/index.js (hunk 1 above)
Comment (scruffian) [inline client/state/purchases/index.js]: Actually, why did you change this from just `
import { action as ActionTypes } from 'lib/upgrades/constants';`?

File: client/state/purchases/index.js [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		predicate = { userId: action.userId };
+	}
+
+	purchases = removeMissingPurchasesByPredicate( existingPurchases, action.purchases, predicate );
+	purchases = overwriteExistingPurchases( purchases, action.purchases );
+
+	return purchases;
+}
+
+export const items = ( state = initialState, action ) => {
```
Comment (scruffian) [inline client/state/purchases/index.js]: Why `items`?

File: client/state/purchases/test/index.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+} from 'state/action-types';
+import { getByPurchaseId } from '../selectors';
+import { items } from '..';
+
+describe( 'items', () => {
+	const userId = 1337,
+		siteId = 2701;
+
+	it( 'should return an object with the initial state', () => {
+		expect( items( undefined, { type: 'UNRELATED' } ) ).to.be.eql( {
```
Comment (scruffian) [inline client/state/purchases/test/index.js]: Would it be a bad idea to import `initialState` from the reducer?

**Comment by scruffian** [comment]
This looks great. I had a few minor questions...

**Comment by drewblaisdell** [comment]
I'm going to update this to use `createReducer` as well.

File: client/state/action-types.js [hunk 5]
```diff
@@ -14,33 +14,35 @@ export const ANALYTICS_EVENT_RECORD = 'ANALYTICS_EVENT_RECORD';
 export const ANALYTICS_MULTI_TRACK = 'ANALYTICS_MULTI_TRACK';
```
Comment (gziolo) [inline client/state/action-types.js]: Did you order those constants manually or do you use anything fancy to do it for you? :)

File: client/state/purchases/index.js (hunk 1 above)
Comment (gziolo) [inline client/state/purchases/index.js]: This is a common pattern used inside `state` folder.

File: client/state/purchases/index.js [hunk 6]
```diff
@@ -0,0 +1,173 @@
+/**
```
Comment (gziolo) [inline client/state/purchases/index.js]: This file should be renamed to `reducer.js` to follow other folders.

File: client/state/purchases/index.js (hunk 3 above)
Comment (gziolo) [inline client/state/purchases/index.js]: I think it should be named `purchases`.

File: client/state/purchases/selectors.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
+// External dependencies
+import assign from 'lodash/assign';
+import find from 'lodash/find';
+
+export const getByPurchaseId = ( state, id ) => (
```
Comment (gziolo) [inline client/state/purchases/selectors.js]: Wouldn't it better to name it `getByPurchaseById`?

**Comment by gziolo** [comment]
> I'm going to update this to use createReducer as well.

It probably can wait. Let's add unit test for a selector and merge it as it as, because it's already quite big PR. Next step would be to extract actions and move action creators. Then we should be good to connect it with reducer tree and replace Flux stores.

File: client/state/action-types.js (hunk 5 above)
Comment (stephanethomas) [inline client/state/action-types.js]: You can use <kbd>F9</kbd/> in Sublime Text to sort lines for you.

Comment (gziolo) [inline client/state/action-types.js]: What about PhpStorm or WebStorm? :D

Comment (drewblaisdell) [inline client/state/action-types.js]: I used `:sort` in vim.

File: client/state/purchases/index.js (hunk 3 above)
Comment (drewblaisdell) [inline client/state/purchases/index.js]: I used `items` because it seems to be a relatively common pattern in the state tree, e.g. [state.posts.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101) and [state.sites.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101).

File: client/state/purchases/index.js (hunk 2 above)
Comment (drewblaisdell) [inline client/state/purchases/index.js]: Awesome.

File: client/state/purchases/test/index.js (hunk 4 above)
Comment (drewblaisdell) [inline client/state/purchases/test/index.js]: I didn't consider doing this because I just copied the tests. @Tug has mentioned that he thinks this is a bad idea but I can't remember why.

File: client/state/purchases/selectors.js (hunk 7 above)
Comment (drewblaisdell) [inline client/state/purchases/selectors.js]: Well, there is only one criteria, so I would expect `by` to appear only once, but I might be misunderstanding the ambiguity of `getByPurchaseId`. The reason I wouldn't use `getPurchaseById` is that you can get purchase by the user or site ID as well.

**Comment by drewblaisdell** [comment]
> It probably can wait. Let's add unit test for a selector and merge it as it as, because it's already quite big PR.

:+1:

I moved the reducer to `reducer.js`, updated it to use the object spread operator, added a test for the selectors, and rebased.

File: client/state/purchases/selectors.js (hunk 7 above)
Comment (gziolo) [inline client/state/purchases/selectors.js]: I missed that first by when checking on mobile, sorry about that :)

**Comment by gziolo** [comment]
LGTM 👍

**Comment by gziolo** [comment]
@drewblaisdell make sure you apply this PR #6313 properly when rebasing with master ;)

**Comment by scruffian** [comment]
@Automattic/sdev-feed

**Comment by drewblaisdell** [comment]
Thanks @gziolo and @scruffian.
